code-studio/scripts/build-commands.js

```diff
@@ -172,39 +182,53 @@ exports.ensureDirectoryExists = function (dir) {
 };
 
 /**
- * Execute a sequence of shell commands as a build script, with useful output
- * wherever possible.
- * Will call process.exit() if the command fails for any reason.
+ * Execute a list of shell commands in sequence.
  * @param {string[]} commands - array of shell commands to be executed in sequence.
+ * @return {Promise} which resolves when all commands complete (to the stdout
+ *         value of the last command in the sequence), or rejects if any of
+ *         them fail.
  */
-exports.execute = function (commands) {
-  commands.forEach(function (command) {
-    try {
-      console.log(command);
-
-      // For documentation on synchronous execution of shell commands from node scripts, see:
-      // https://nodejs.org/docs/latest/api/child_process.html#child_process_synchronous_process_creation
-      var result = child_process.execSync(command, {
-        env: _.extend({}, process.env, {
-          PATH: './node_modules/.bin:' + process.env.PATH
-        }),
-        stdio: 'inherit'
+exports.executeSequence = function (commands) {
+  var options = getChildProcessOptions();
+  return new Promise(function (resolve, reject) {
+    var runNextCommand = function (commands) {
+      var nextCommand = commands[0];
+      console.log(nextCommand);
+      child_process.exec(nextCommand, options, function (error, stdout) {
+        if (error) {
+          reject(transformExecError(error));
+        } else if (1 === commands.length) {
+          resolve(stdout);
+        } else {
+          runNextCommand(commands.slice(1));
+        }
       });
-    } catch (e) {
-      exports.exitWithError(e);
-    }
+    };
+    runNextCommand(commands);
   });
 };
```

Comment:
This feels a little clunky. There are libraries we could use that make this sort of serial chain of promises easier to handle. Not sure it's worth it for one case. Here's an (untested) attempt I had of making it better. I'm not sure I really succeeded
```
// Execute a single command, returns a promise
function executeCommand(command, options) {
  return new Promise(function (resolve, reject) {
    child_process.exec(command, options, function (error, stdout) {
      if (error) {
        reject(transformExecError(error));
      } else {
        console.log(stdout);
        resolve();
      }
    });
  });
}

// Converts commands into a chained set of promises
exports.executeSequence = function (commands) {
  var options = getChildProcessOptions();
  var promises = commands.map(function (command) {
    return executeCommand(command, options);
  });

  // if you wanted to be even more functional, you could probably create a promise chain using commands.reduce instead of using .map and a for loop.
  var promiseChain = promises[0];
  for (var i = 1; i < promises.length; i++) {
    promiseChain.then(promises[i]);
  }

  return promiseChain;
};